src/topology/ping_pong.rs

```diff
@@ -0,0 +1,865 @@
+/// VDAF describes `Start` and `Rejected` states, but the `Start` state is never instantiated in
+/// code, and the `Rejected` state is represented as `std::result::Result::Err`, so this enum does
+/// not include those variants.
+///
+/// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum State<
+    const VERIFY_KEY_SIZE: usize,
+    const NONCE_SIZE: usize,
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+> {
+    /// Preparation of the report will continue with the enclosed state.
+    Continued(A::PrepareState),
+    /// Preparation of the report is finished and has yielded the enclosed output share.
+    Finished(A::OutputShare),
+}
+
+/// Values returned by [`PingPongTopology::continued`].
+#[derive(Clone, Debug)]
+pub enum ContinuedValue<
+    const VERIFY_KEY_SIZE: usize,
+    const NONCE_SIZE: usize,
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+> {
+    /// The operation resulted in a new state and a message to transmit to the peer.
+    WithMessage {
+        /// The transition that will be executed. Call `Transition::evaluate` to obtain the next
+        /// [`State`] and a [`Message`] to transmit to the peer.
+        transition: Transition<VERIFY_KEY_SIZE, NONCE_SIZE, A>,
+    },
+    /// The operation caused the host to finish preparation of the input share, yielding an output
+    /// share and no message for the peer.
+    FinishedNoMessage {
+        /// The output share which may now be accumulated.
+        output_share: A::OutputShare,
+    },
+}
+
+/// Extension trait on [`crate::vdaf::Aggregator`] which adds the [VDAF Ping-Pong Topology][VDAF].
+///
+/// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+pub trait PingPongTopology<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>:
+    Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>
+{
+    /// Specialization of [`State`] for this VDAF.
+    type State;
+    /// Specialization of [`ContinuedValue`] for this VDAF.
+    type ContinuedValue;
+    /// Specializaton of [`Transition`] for this VDAF.
+    type Transition;
+
+    /// Initialize leader state using the leader's input share. Corresponds to
+    /// `ping_pong_leader_init` in the forthcoming `draft-irtf-cfrg-vdaf-07`.
+    ///
+    /// If successful, the returned [`Message`] (which will always be `Message::Initialize`) should
+    /// be transmitted to the helper. The returned [`State`] (which will always be
+    /// `State::Continued`) should be used by the leader along with the next [`Message`] received
+    /// from the helper as input to [`Self::continued`] to advance to the next round.
+    fn leader_initialize(
+        &self,
+        verify_key: &[u8; VERIFY_KEY_SIZE],
+        agg_param: &Self::AggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &Self::PublicShare,
+        input_share: &Self::InputShare,
+    ) -> Result<(Self::State, Message), PingPongError>;
```

Comment:
Why do we only define the method here, then implement it for `Aggregator` below? Couldn't we just make these derived methods?

Reply:
The alternative would be to provide default methods inside the extension trait definition. This might be feasible, since the trait definition and the blanket implementation have the same bounds on their generic parameters. Note that the body of the blanket implementation block would then just have associated type definitions. Six of one, half a dozen of the other, I think.

Reply:
I'm not sure I understand David's explanation here. To be clear, instead of

```rust
trait PingPong {
  fn cool();
}

impl PingPong for Vdaf {
  fn cool() {
    println!("guy");
  }
}
```

I was suggesting something like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021

This seems a bit cleaner to me, but I'm happy to defer to you guys.

Reply:
Whoopsies, here's a link to the playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7a9e4dd238dab4056dbee25312e3068c

Reply:
OK, so unpacking Chris' suggestion, there's two implications here.

One is that we could hoist the implementations of the `PingPongTopology` methods into the trait declaration instead of providing them in the blanket implementation. I think all that does is move some code around in the file and makes no observable difference to users of the crate (except, see next point).

The other implication is that if we remove the blanket impl of `PingPongTopology`, then individual VDAFs like `Prio3` and `Poplar1` have to opt into `PingPongTopology` by doing `impl PingPongTopology<generics> for MyCoolVdaf {}`.

So I think what we have to decide is:

Should VDAFs get ping-pong topology support automatically, for free, just by virtue of being VDAFs, **or** should a VDAF be required to explicitly opt into ping-pong (via an empty impl block)?

My 2c: the intent of the ping-pong topology in the spec is that it works for any VDAF, so a blanket implementation is appropriate. Let's also consider a future where there are VDAFs implemented outside of crate `prio`. If we add a new topology to `prio`, then it's not usable until each downstream crate that implements a VDAF does a release just to add an empty impl block. I think that just introduces extra delay and release chores for something like a DAP implementation that wants to use a new topology.

I think we should use the current implementation, because given the problem we have right now (one topology across two VDAFs, three if you count the dummy VDAF), then going one way or another makes no difference to users of the crate.

Reply:
One further reason to keep the blanket impl: if there's no blanket impl, and the trait provides default implementations for each ping pong method, then other VDAFs could override the default methods with their own code when implementing the PingPongTopology trait, which is unnecessary flexibility and opens up the door for mistakes.

Assuming we still have a blanket impl, I agree the choice of where to put the code doesn't matter much.